Add GlobalInit to blocks to emit initialization instructions for globals

gysddn opened the pull request:
`GlobalInit` block is now static block within a `SemIR` which will be used to emit initialization instructions for variables in the `Package` scope.

inst_block_stack now has additional methods to handle `GlobalInit` block separately, this block can be popped without being finalized allowing to accumulate between all instances of variables.

At the end of the `check` phase, if this block is not empty , the function `__global_init` will be added with this block being inserted into it.

This block is pushed to `inst_block_scope` at the end `BindName`, allowing instruction to be emitted into it, then popped at the semicolon (VariableDecl).

This significantly changes the `SemIR` output, that's why this commit updates a lot of the test cases.

gysddn committed:
Add Inits to blocks to emit initialization insructions for globals

`Inits` block is now static block within a `SemIR` which will be used to
emit initialization instructions for variables in the `Package` scope.

inst_block_stack now has additional methods to handle `Inits` block
separately, this block can be popped without being finalized allowing to
accumulate between all instances of variables.

At the end of the `check` phase, if this block is not empty , the
function `__global_init` will be added with this block being inserted
into it.

This significantly changes the `SemIR` output, that's why this commit
updates a lot of the test cases.

gysddn commented:
Only implemented for 'var', not 'let' at the moment.

geoffromer commented on toolchain/check/handle_binding_pattern.cpp:
```diff
@@ -115,6 +115,10 @@ auto HandleAnyBindingPattern(Context& context, Parse::NodeId parse_node,
         RegisterReturnedVar(context, bind_id);
       }
+      if (auto current_scope_id = context.current_scope_id();
+          current_scope_id == SemIR::NameScopeId::Package) {
```
The named variable doesn't seem necessary here. Would this work instead?

```suggestion
      if (context.current_scope_id() == SemIR::NameScopeId::Package) {
```

gysddn replied:
Definitely! This totally has slipped my attention.
I updated them.

geoffromer commented on toolchain/check/handle_variable.cpp:
```diff
@@ -95,6 +95,10 @@ auto HandleVariableDecl(Context& context, Parse::VariableDeclId parse_node)
 
   context.decl_state_stack().Pop(DeclState::Var);
+  if (auto current_scope_id = context.current_scope_id();
+      current_scope_id == SemIR::NameScopeId::Package) {
```
```suggestion
  if (context.current_scope_id() == SemIR::NameScopeId::Package) {
```

geoffromer commented on toolchain/check/inst_block_stack.h:
```diff
@@ -29,6 +29,8 @@ class InstBlockStack {
   // called in order to support lazy allocation.
   auto Push() -> void { Push(SemIR::InstBlockId::Invalid); }
 
+  auto PushInit() -> void;
```
Could you add comments to this and `PopInit`? It'd be great to have some of the information from the PR description in the code itself.

gysddn replied:
Done!

jonmeow commented on toolchain/check/context.cpp:
```diff
@@ -618,6 +618,28 @@ auto Context::ParamOrArgEnd(Parse::NodeKind start_kind) -> SemIR::InstBlockId {
+  inst_block_stack().PushInit();
+  if (!inst_block_stack().PeekCurrentBlockContents().empty()) {
+    AddInst({Parse::NodeId::Invalid, SemIR::Return{}});
+    // Pop the Inits block here to finalize it
```
```suggestion
    // Pop the Inits block here to finalize it.
```

Style nit: we use [standard punctuation](https://google.github.io/styleguide/cppguide.html#Punctuation,_Spelling_and_Grammar) in comments.

jonmeow commented on toolchain/check/context.cpp:
```diff
@@ -618,6 +618,28 @@ auto Context::ParamOrArgEnd(Parse::NodeKind start_kind) -> SemIR::InstBlockId {
+    // Pop the Inits block here to finalize it
+    inst_block_stack().Pop();
+
+    // __global_init is only added if there are initialization instructions
```
```suggestion
    // __global_init is only added if there are initialization instructions.
```

Style: punctuation

jonmeow commented on toolchain/check/context.cpp:
```diff
@@ -618,6 +618,28 @@ auto Context::ParamOrArgEnd(Parse::NodeKind start_kind) -> SemIR::InstBlockId {
+    // __global_init is only added if there are initialization instructions
+    auto nameid = sem_ir().identifiers().Add("__global_init");
+    sem_ir().functions().Add(
+        {.name_id = SemIR::NameId::ForIdentifier(nameid),
```
```suggestion
    auto name_id = sem_ir().identifiers().Add("__global_init");
    sem_ir().functions().Add(
        {.name_id = SemIR::NameId::ForIdentifier(name_id),
```

Style: underscores are used to separate words in [variable names](https://google.github.io/styleguide/cppguide.html#Variable_Names).

geoffromer commented on toolchain/check/inst_block_stack.cpp:
```diff
@@ -21,6 +21,11 @@ auto InstBlockStack::Push(SemIR::InstBlockId id) -> void {
 
+auto InstBlockStack::PushInit() -> void {
+  Push(SemIR::InstBlockId::Inits);
+  stack_[size_ - 1].content = init_block_;
```
Is it possible for`size_` to not be 0 here? I'd naively expect initializers to always be at the bottom of the stack. If so, maybe rather than pushing and popping them, we could just always have them at the bottom of the stack?

jonmeow replied:
I believe the package's inst block is always at the bottom of the stack. If you consider something like a global `let x: MyType(1 + 7) = Foo;`, the `let x: MyType(1 + 7)` goes into the package's inst block, while the `= Foo;` is going into the global init block. That's going to inhibit leaving global inits on the bottom of the stack.

jonmeow commented on toolchain/check/context.cpp:
```diff
@@ -618,6 +618,28 @@ auto Context::ParamOrArgEnd(Parse::NodeKind start_kind) -> SemIR::InstBlockId {
 }
 
+auto Context::FinalizeInits() -> void {
+  inst_block_stack().PushInit();
+  if (!inst_block_stack().PeekCurrentBlockContents().empty()) {
```
Had you considered manipulating `inits_` directly here, rather than adding it to the block stack?

gysddn replied:
I did, but I just couldn't come up with a minimal way to do it.

The `GlobalInit` is being constructed inside the stack object, so there is no direct way to either "view" or "manipulate" it. I had to check if it was empty or not, and if it wasn't, I had to push it onto the stack again to add the `Return` instruction. I just thought I would push it regardless, that way I would be able to both "check its emptiness" and "add the return".

I realize it is not the most straight-forward, but I didn't want to make any more API changes to `InstBlockStack` than I had to.

jonmeow replied:
Makes sense, with the `return` function. Makes me wonder if we should switch around how we organize `Finalize` logic, now that we're adding more (if we had a single entry point for both exports and __global_init then it might look a little more reasonable to put this full function on InstBlockStack), but this seems good for now.

gysddn replied:
That makes sense. I might try that and see how it would work.

jonmeow commented on toolchain/sem_ir/ids.h:
```diff
@@ -362,6 +362,11 @@ struct InstBlockId : public IdBase, public Printable<InstBlockId> {
+  // Global declaration initialization instructions. Empty if none are present.
+  // Otherwise, __global_init function will be generated and this block will
+  // be inserted into it.
+  static const InstBlockId Inits;
```
Why do you choose `Inits` versus something such as `GlobalInit` to match the produced function? It may be helpful to be more specific in order to disambiguate from local initializers.

gysddn replied:
I agree.
I renamed the block to `GlobalInit`.

jonmeow commented on toolchain/check/context.h:
```diff
@@ -363,6 +363,9 @@ class Context {
     inst_blocks().Set(SemIR::InstBlockId::Exports, exports_);
   }
 
+  // Finalizes the initialization function (__global_init)
```
```suggestion
  // Finalizes the initialization function (__global_init).
```

Style: Punctuation.

gysddn replied:
I updated them all.

jonmeow commented on toolchain/check/inst_block_stack.cpp:
```diff
@@ -52,6 +57,14 @@ auto InstBlockStack::Pop() -> SemIR::InstBlockId {
+  CARBON_CHECK(stack_[size_ - 1].id == SemIR::InstBlockId::Inits)
+      << "Trying to pop Inits block from " << name_
+      << " but a different block is present!";
+  init_block_ = stack_[size_ - 1].content;
```
```suggestion
  init_block_ = std::move(stack_[size_ - 1].content);
```

jonmeow commented on toolchain/check/inst_block_stack.cpp:
```diff
@@ -21,6 +21,11 @@ auto InstBlockStack::Push(SemIR::InstBlockId id) -> void {
 
+auto InstBlockStack::PushInit() -> void {
+  Push(SemIR::InstBlockId::Inits);
+  stack_[size_ - 1].content = init_block_;
```
```suggestion
  stack_[size_ - 1].content = std::move(init_block_);
```

At present, this would perform a copy on PushInit and PopInit. That would create scaling issues when there are many or complex initializers. Using `move` should alleviate the issue.

gysddn replied:
Done!

jonmeow reviewed (commented):
Thanks, the output looks good! My comments are mostly style-related, this approach looks right to me.

gysddn committed:
Fix some code style problems.

- Add doc comments for `PushInit` and `PopInit` methods
- Fix punctuation in some places.
- Remove redundant named variables in `if`s

gysddn committed:
move Inits block contents instead of copying them each time

- updates `PopInit` and `PushInit` functions to move the contents of
Inits block

gysddn committed:
Rename `Inits` inst block to `GlobalInit`

- Rename the InstBlockId enum
- Update SemIR output text from 'inits' to 'global_init'
- Update tests

jonmeow commented on toolchain/check/context.h:
```diff
@@ -363,6 +363,9 @@ class Context {
   }
 
+  // Finalizes the initialization function (__global_init).
+  auto FinalizeInits() -> void;
```
For FinalizeInits/PushInit/PopInit, could you please rename them using Globalnit similar to the block rename?

gysddn replied:
Done.

gysddn committed:
Rename `*Inits` block functions to `*GlobalInit`

jonmeow reviewed (approved):
Looks good, thanks!

jonmeow commented:
Looking at https://github.com/carbon-language/carbon-lang/actions/runs/7833531743/job/21374710204, I think the PR needs to pull in changes made at head. Specifically #3688 factored out current_scope_id and related functionality to its own class.

gysddn committed:
Merge branch 'trunk' into check-global

This commit updates #3687 to fix conflicts.

- Update scope check expressions, and use the new ScopeStack.
- Fix conflicts in the test files.
- Update new test files.

gysddn commented:
I added to merge commit, but if that seems too excessive to review. I can rebase this branch and make it more minimal.
